### ISSUE TITLE
Cache entire pages per customer group

### ISSUE THREAD

**Issue by mikeschaekermann** [nexcess/magento-turpentine#964]
Is it possible to cache all category pages and product pages per customer group?

We use the Netzarbeiter/GroupCatalogs2 extension (see https://github.com/Vinai/groupscatalog2) to restrict access to certain category pages. If guest users (i.e., not logged in) try to access a category, they will be 302-redirected to /customer/account/login. Therefore, the cache should differentiate between logged-in and non-logged-in users when rendering category and product pages.

Is this possible? Any help would be greatly appreciated!

**Comment by miguelbalparda** [comment]
Please see #465 and #665. #838 seems to be what you are looking for.

**Comment by mikeschaekermann** [comment]
Thanks! I noticed that there is already the `<access>customer_group</access>` parameter available for ESI Cache Policies. However, you can only use it for hole-punching directives for certain blocks, not for entire page types as far as I could see. Please correct me if I am wrong.

If I follow the approach, suggested in #665 with:

```
if (req.http.X-User-Group) {
    hash_data("usergroup="+req.http.X-User-Group);
}
```

I don't see a way to apply this logic to certain page types only because I cannot differentiate between category or product pages and other page types just from the URL.

Could you give me a hint if you see a way to do this? Thank you for your time! :)

**Comment by miguelbalparda** [comment]
You might want to see #942 in conjunction with the customer_group policy.

**Comment by mikeschaekermann** [comment]
Thanks @miguelbalparda. I ended up caching the entire website per customer group by changing:

```
    if (req.http.X-Varnish-Esi-Access == "customer_group" &&
            req.http.Cookie ~ "customer_group=") {
        hash_data(regsub(req.http.Cookie, "^.*?customer_group=([^;]*);*.*$", "\1"));
    }
```

in my Varnish 3.0 VCL file to:

```
    if (req.http.Cookie ~ "customer_group=") {
        hash_data(regsub(req.http.Cookie, "^.*?customer_group=([^;]*);*.*$", "\1"));
    }
```

This way, every request will be cached on a per-customer-group basis, not only categories and product pages. We do not have a consistent URL pattern to recognize those page types. So, for now, I am going with the more general solution of caching every request per customer group. Thanks for your help!

**Comment by ADDISON74** [comment]
Will be this included in next versions?

**Comment by miguelbalparda** [comment]
There are no PRs for this so far and we think the current solution is working as expected. New comers can see this issue with the proposed solution if they face the same situation @ADDISON74. If you think this is something we should include feel free to create a PR.

**Comment by colinodell** [comment]
I also need the ability to cache certain pages by customer group.  For example, my price blocks render differently depending on whether or not you have a membership:

![2015-10-08_14-47-37](https://cloud.githubusercontent.com/assets/202034/10376385/aeae7774-6dcb-11e5-83ab-cdf4d51dedd7.png)

![2015-10-08_14-48-17](https://cloud.githubusercontent.com/assets/202034/10376384/aeaafa90-6dcb-11e5-9ba6-9103881d0417.png)

@ADDISON74's solution works great for me.  Perhaps this could be a toggleable config option?

The downside is that EVERYTHING gets cached per-group, when I only need this on a sub-section of pages.  An alternate implementation could be something like this:

~~~xml
<some_layout_handle>
    <reference name="root">
        <action method="setCacheOptions">
            <params>
                <access>customer_group</access>
            </params>
            </action>
    </reference>
</some_layout_handle>
~~~

Basically a slimmed-down version of `setEsiOptions` which only works on the `root` element and only supports the `access` param, at least to start with.  You can then set a `X-Varnish-Access` header and check that.

**Comment by colinodell** [comment]
For example, you could throw something like this near the top of `Nexcessnet_Turpentine_Model_Observer_Esi::injectEsi`:

~~~php
if ($blockObject->getNameInLayout() === 'root' && $options = $blockObject->getVarnishOptions()) {
    if (isset($options['access']) && $options['access'] === 'customer_group') {
        Mage::app()->getResponse()->setHeader('X-Varnish-Cache-Group', Mage::getSingleton('customer/session')->getCustomerGroupId());
    }
}
~~~

I haven't tested this, and it's probably not the best location, but it should probably work.

**Comment by puecher** [comment]
Hi @mikeschaekermann,

I have the following issue using your approach for Varnish 4.0:

https://github.com/nexcess/magento-turpentine/issues/1051

Does it work for you?

**Comment by mikeschaekermann** [comment]
@puecher I haven't had the chance to test it with Varnish 4.0 yet, sorry!
